src/shogun/machine/GLM.cpp

```diff
@@ -0,0 +1,130 @@
+#include <shogun/optimization/GradientDescendUpdater.h>
+#include <shogun/optimization/ConstLearningRate.h>
+
+#include<iostream>
+// #include <utility>
+#include <cmath>
+
+using namespace shogun;
+
+
+
+
+
+GLM::GLM()
+{
+	std::cout<<"Entered non-parameterized constructor.\n";
+	SG_ADD(&distribution, "distribution_type", "variable to store name of distribution type", ParameterProperties::HYPER);
+	SG_ADD(&m_eta, "eta", "threshold parameter that linearizes the exp() function above eta", ParameterProperties::HYPER);
+	SG_ADD(&m_lambda, "lambda", "regularization parameter of penalty term", ParameterProperties::HYPER);
+	SG_ADD(&m_alpha, "alpha", "weighting between L1 penalty and L2 penalty term of the loss function", ParameterProperties::HYPER);
+	SG_ADD(&m_tolerance, "tolerance", "convergence threshold or stopping criteria", ParameterProperties::HYPER);
+	
+	m_gradient_updater = std::make_shared<GradientDescendUpdater>();
+	m_learning_rate = std::make_shared<ConstLearningRate>();
+	m_penalty = std::make_shared<ElasticNetPenalty>();
+	m_cost_function = std::make_shared<GLMCostFunction>();
+
+}
+
+GLM::GLM(GLM_DISTRIBUTION distr, float64_t alpha, float64_t lambda, float64_t learning_rate, int32_t max_iterations, float64_t tolerance, float64_t eta): RandomMixin<IterativeMachine<LinearMachine>>()
+{
+	std::cout<<"Entered parameterized constructor.\n";
+	distribution=distr;
+	m_alpha=alpha;
+	m_lambda=lambda;
+	m_max_iterations=max_iterations;
+	m_tolerance=tolerance;
+	m_eta=eta;
+
+	m_learning_rate->set_const_learning_rate(learning_rate);
+
+	m_penalty->set_l1_ratio(m_alpha);
+
+	m_cost_function->set_target(shared_from_this()->as<GLM>());
+}
+
+std::shared_ptr<RegressionLabels> GLM::apply_regression(std::shared_ptr<Features> data)
+{
+	LinearMachine::set_features(std::static_pointer_cast<DotFeatures>(data));
+
+	if (!LinearMachine::features)
+		return std::make_shared<RegressionLabels>(SGVector<float64_t>());
+
+	auto num = LinearMachine::features->get_num_vectors();
+	ASSERT(num>0)
+	ASSERT(m_w.vlen==features->get_dim_feature_space())
+	SGVector<float64_t> out(num);
+	LinearMachine::features->dense_dot_range(out.vector, 0, num, NULL, m_w.vector, m_w.vlen, bias);
+	auto result = m_cost_function->non_linearity(out);
+	return std::make_shared<RegressionLabels>(result);
+}
+
+void GLM::init_model(const std::shared_ptr<Features>& data)
+{
+	ASSERT(m_labels)
+	if (data)
+	{
+		if (!data->has_property(FP_DOT))
+			error("Specified features are not of type CDotFeatures");
+		LinearMachine::set_features(std::static_pointer_cast<DotFeatures>(data));
+	}
+	ASSERT(features)
+
+	NormalDistribution<float64_t> normal_dist;
+	auto n_features = LinearMachine::features->get_dim_feature_space();
+	
+	if (m_compute_bias && bias == 0)
+		bias = 1 / (n_features + 1) * normal_dist(m_prng);
+
+	if(m_w.size() > 0)
+	{
+		m_w = SGVector<float64_t>(n_features);
+		for (auto& weight : m_w)
+			weight = 1 / (n_features + 1) * normal_dist(m_prng);
```

Comment:
This `normal_dist(m_prng)` is always returning Zero. Why might that be happening? Is there any workaround?

Reply:
You are dividing 1, an int, by an int larger than 1. You are expecting a double but you’re getting a fraction as an int, in this case will always be 0. You need to write 1. for it to be a double

Reply:
Thanks.